File: 2019/day12.go
package main

import (
	"github.com/lukechampine/advent/utils"
)

var input = utils.ReadInput("day12_input.txt")

type Vec3 struct {
	X, Y, Z int
}

func (v Vec3) energy() int {
	return utils.Abs(v.X) + utils.Abs(v.Y) + utils.Abs(v.Z)
}

type Moon struct {
	Pos Vec3
	Vel Vec3
}

func (m *Moon) energy() int {
	return m.Pos.energy() * m.Vel.energy()
}

func (m *Moon) gravitate(n Moon) {
	adjust := func(a, b int) int {
		if a == b {
			return 0
		} else if a < b {
			return 1
		} else {
			return -1
		}
	}
	m.Vel.X += adjust(m.Pos.X, n.Pos.X)
	m.Vel.Y += adjust(m.Pos.Y, n.Pos.Y)
	m.Vel.Z += adjust(m.Pos.Z, n.Pos.Z)
}

func (m *Moon) move() {
	m.Pos.X += m.Vel.X
	m.Pos.Y += m.Vel.Y
	m.Pos.Z += m.Vel.Z
}

func main() {
	// part 1
	var positions []Vec3
	utils.Parse(&positions, "<x=%d, y=%d, z=%d>", input)
	moons := make([]Moon, len(positions))
	for i := range moons {
		moons[i].Pos = positions[i]
	}

	for step := 0; step < 1000; step++ {
		for i := range moons {
			for j := range moons {
				if i == j {
					continue
				}
				moons[i].gravitate(moons[j])
			}
		}
		for i := range moons {
			moons[i].move()
		}
	}

	utils.Println(utils.Sum(len(moons), func(i int) int {
		return moons[i].energy()
	}))

	// part 2
	for i := range moons {
		moons[i] = Moon{
			Pos: positions[i],
		}
	}
	var hist [3]map[[8]int]int
	for i := range hist {
		hist[i] = make(map[[8]int]int)
	}
	var cycles Vec3
	for step := 0; cycles.X*cycles.Y*cycles.Z == 0; step++ {
		mxs := [8]int{
			moons[0].Pos.X, moons[1].Pos.X, moons[2].Pos.X, moons[3].Pos.X,
			moons[0].Vel.X, moons[1].Vel.X, moons[2].Vel.X, moons[3].Vel.X,
		}
		mys := [8]int{
			moons[0].Pos.Y, moons[1].Pos.Y, moons[2].Pos.Y, moons[3].Pos.Y,
			moons[0].Vel.Y, moons[1].Vel.Y, moons[2].Vel.Y, moons[3].Vel.Y,
		}
		mzs := [8]int{
			moons[0].Pos.Z, moons[1].Pos.Z, moons[2].Pos.Z, moons[3].Pos.Z,
			moons[0].Vel.Z, moons[1].Vel.Z, moons[2].Vel.Z, moons[3].Vel.Z,
		}
		if cycles.X == 0 {
			if oldstep, ok := hist[0][mxs]; ok {
				cycles.X = step - oldstep
			}
			hist[0][mxs] = step
		}
		if cycles.Y == 0 {
			if oldstep, ok := hist[1][mys]; ok {
				cycles.Y = step - oldstep
			}
			hist[1][mys] = step
		}
		if cycles.Z == 0 {
			if oldstep, ok := hist[2][mzs]; ok {
				cycles.Z = step - oldstep
			}
			hist[2][mzs] = step
		}

		for i := range moons {
			for j := range moons {
				if i == j {
					continue
				}
				moons[i].gravitate(moons[j])
			}
		}
		for i := range moons {
			moons[i].move()
		}
	}
	utils.Println(utils.LCM(cycles.X, cycles.Y, cycles.Z))
}

File: 2019/day12.zig
const std = @import("std");
const utils = @import("utils.zig");

const Vec3 = struct {
    x: i64 = 0,
    y: i64 = 0,
    z: i64 = 0,

    fn energy(s: Vec3) i64 {
        var sum: i64 = 0;
        inline for (std.meta.fields(Vec3)) |f| {
            sum += utils.abs(@field(s, f.name));
        }
        return sum;
    }
};

const Moon = struct {
    pos: Vec3 = Vec3{},
    vel: Vec3 = Vec3{},

    fn energy(m: Moon) i64 {
        return m.pos.energy() * m.vel.energy();
    }

    fn move(m: *Moon) void {
        inline for (std.meta.fields(Vec3)) |f| {
            @field(m.pos, f.name) += @field(m.vel, f.name);
        }
    }

    fn gravitate(m: *Moon, n: Moon) void {
        inline for (std.meta.fields(Vec3)) |f| {
            @field(m.vel, f.name) += utils.sign(i64, @field(n.pos, f.name) - @field(m.pos, f.name));
        }
    }
};

pub fn main() void {
    const input = utils.readFileLines("day12_input.txt");
    var initPositions: [4]Vec3 = undefined;
    for (initPositions) |*p, i| {
        var coords = utils.splitByte(input[i][0 .. input[i].len - 1], ',');
        inline for (std.meta.fields(Vec3)) |f, j| {
            @field(p, f.name) = utils.parseInt(i64, coords[j][3..]);
        }
    }

    // part 1
    var moons: [4]Moon = undefined;
    for (moons) |*m, i| {
        m.pos = initPositions[i];
        m.vel = Vec3{};
    }
    var steps: usize = 0;
    while (steps < 1000) : (steps += 1) {
        for (moons) |*m, i| {
            for (moons) |n, j| {
                if (i == j) continue;
                m.gravitate(n);
            }
        }
        for (moons) |*m| {
            m.move();
        }
    }
    var energy: i64 = 0;
    for (moons) |m| {
        energy += m.energy();
    }
    utils.println(energy);

    // part 2
    var initialMoons = [4]Moon{
        Moon{ .pos = Vec3{ .x = 17, .y = -7, .z = -11 } },
        Moon{ .pos = Vec3{ .x = 1, .y = 4, .z = -1 } },
        Moon{ .pos = Vec3{ .x = 6, .y = -2, .z = -6 } },
        Moon{ .pos = Vec3{ .x = 19, .y = 11, .z = 9 } },
    };
    moons = initialMoons;
    steps = 0;
    var cycle = Vec3{};
    while (cycle.x * cycle.y * cycle.z == 0) : (steps += 1) {
        inline for (std.meta.fields(Vec3)) |cf| {
            if (@field(cycle, cf.name) == 0) {
                var match: bool = true;
                for (initialMoons) |m, i| {
                    match = match and @field(m.pos, cf.name) == @field(moons[i].pos, cf.name);
                    match = match and @field(m.vel, cf.name) == @field(moons[i].vel, cf.name);
                }
                if (match) {
                    @field(cycle, cf.name) = @intCast(i64, steps);
                }
            }
        }

        for (moons) |*m, i| {
            for (moons) |n, j| {
                if (i == j) continue;
                m.gravitate(n);
            }
        }
        for (moons) |*m| {
            m.move();
        }
    }
    utils.println(utils.lcm(cycle.x, utils.lcm(cycle.y, cycle.z)));
}
